### PR TITLE
Don't confuse 10-char folder with hash

### PR DESCRIPTION
**Pull request by maxime-rainville**
This PR tweaks the logic that resolve file ID to make sure that when the hash is found from the file ID, it is matches the actual hash of the file.

# Parent issue 
https://github.com/silverstripe/silverstripe-assets/issues/269

### CODE REVIEW

**Comment by maxime-rainville** [comment]
A thing to note is that this will slow down some request because we'll have to recompile the hash of files pretty much every time we want to serve a protected file.

**Comment by dnsl48** [comment]
>  we'll have to recompile the hash of files pretty much every time 

why don't we keep once precalculated hash in the filename by the variant?

File: src/FilenameParsing/HashFileIDHelper.php [hunk 1]
```diff
@@ -79,7 +79,7 @@ public function cleanFilename($filename)
 
     public function parseFileID($fileID)
     {
-        $pattern = '#^(?<folder>([^/]+/)*)(?<hash>[a-zA-Z0-9]{10})/(?<basename>((?<!__)[^/.])+)(__(?<variant>[^.]+))?(?<extension>(\..+)*)$#';
+        $pattern = '#^(?<folder>([^/]+/)*)(?<hash>[a-f0-9]{10})/(?<basename>((?<!__)[^/.])+)(__(?<variant>[^.]+))?(?<extension>(\..+)*)$#';
```
Comment (dnsl48) [inline src/FilenameParsing/HashFileIDHelper.php]: if we kept hash in the file ID by the variant, we could make the pattern even narrower, by comparing the hash through a back reference. E.g.:
```php
'#^(?<folder>([^/]+/)*)(?<hash>[a-f0-9]{10})/(?<basename>((?<!__)[^/.])+)(?:__\1)(__(?<variant>[^.]+))?(?<extension>(\..+)*)$#'
```

**Comment by maxime-rainville** [comment]
> @maxime-rainville why don't we keep once pre-calculated hash in the filename by the variant?

The core issue I'm trying to fix here is that you could in some context the folder name can be confused for an hash. 

e.g.: Let's say you have the following path: `decade1980/document.doc`. In this case, `decade1980` could be a valid hash but it could also be a folder name.

I was trying to fix this for file migration, but this could also affect other operations.

**Comment by dnsl48** [comment]
> e.g.: Let's say you have the following path: decade1980/document.doc. In this case, decade1980 could be a valid hash but it could also be a folder name.

yes, understood. I'm thinking that we wouldn't have to calculate the file hash every time if it had it precalculated in the name.
Might not worth the effort, but the idea is
```
decade1980/document__decade1980.doc  - looks suspicious, so we can recalculate its hash to check if it's actually protected file

decade1980/document__aoeu.doc - hashes don't match, don't have to recalculate hash

decade1980/document.doc - doesn't have cache, not protected, don't have to recalculate its hash
```

**Comment by maxime-rainville** [comment]
That wouldn't be possible to achieve in time for the 4.4.0 release. The file hash caching PR could help here ... I've only enabled caching for migration, because I didn't want to wrestle with everything thing else at this time, but we could try enabling in 4.5.0

**Pull request merged**